Preferences / Effect widgets: use '---' placeholder instead of 'None'

ronso0 opened the pull request:
... because in the Sound I/O tabs and in skins' effect regions ` --- ` is much easier to spot/recognize than `None`.

https://bugs.launchpad.net/mixxx/+bug/1892048

ronso0 committed:
Effect widgets/EQ preferences: display '---' when no effect is loaded

ronso0 committed:
Preferences > Sound Hardware: display '---' for 'No sound API'

ronso0 committed:
Preferences > Sound Hardware: display '---' when no sound device is selected

ronso0 commented on src/widget/weffectselector.cpp:
```diff
@@ -72,7 +72,7 @@ void WEffectSelector::populate() {
 
     //: Displayed when no effect is loaded
-    addItem(tr("None"), QVariant());
+    addItem(MIXXX_NOEFFECT_STRING, QVariant());
```
Throughout Mixxx different strings are used as placeholders: `QVariant(QString(""))`, `QString("")` or `QVariant()`
What's safe/correct?

daschuer replied:
https://doc.qt.io/qt-5/qcombobox.html#addItem

I think you can just omit the last parameter and it defaults to QVariant() if the user data is not uses. 
userData is just a container that can store any extra data along with the string.

daschuer commented on src/effects/effectsmanager.h:
```diff
@@ -21,6 +21,8 @@ class EffectChainManager;
 class EffectManifest;
 class EffectsBackend;
 
+#define MIXXX_NOEFFECT_STRING " --- "
```
```suggestion
namespace {
const QString kNoEffectString = QStringLiteral(" --- ");
} // namespace
```

ronso0 replied:
:facepalm: tried that in the cpp before...not in the header

Holzhaus replied:
Yes, anonymous namespace do not belong in header files.

ronso0 replied:
so, should I stick with the pevious MIXXX_NOEFFECT_STRING or what do you propose?

daschuer replied:
Oh I did not notice that this is a header file. 
We should avoid type less macros, the pollute the global name space. 
It can become a static member variable of the effectsmanager. This should be than initialized in the CPP file.

daschuer replied:
like: https://github.com/mixxxdj/mixxx/blob/3085bbe95107fd789a8a966dfeb341cfe74cfad8/src/control/controlbehavior.cpp#L290

ronso0 replied:
yes, I already decided to go with that, like https://github.com/mixxxdj/mixxx/blob/2.3/src/soundio/soundmanagerconfig.cpp#L14

daschuer commented on src/preferences/dialog/dlgprefsound.cpp:
```diff
@@ -45,7 +45,7 @@ DlgPrefSound::DlgPrefSound(QWidget* pParent, SoundManager* pSoundManager,
 
     apiComboBox->clear();
-    apiComboBox->addItem(tr("None"), "None");
+    apiComboBox->addItem(MIXXX_NOSOUNDAPI_STRING, tr("None"));
```
Here "None" as user data is not translated, because it can than be identified in any language.

ronso0 replied:
fixed, it's just `SoundManagerConfig::kEmptyComboBox` now // " --- "

daschuer commented on src/preferences/dialog/dlgprefsound.cpp:
```diff
@@ -621,7 +621,7 @@ void DlgPrefSound::updateAudioBufferSizes(int sampleRateIndex) {
  */
 void DlgPrefSound::refreshDevices() {
-    if (m_config.getAPI() == "None") {
+    if (m_config.getAPI() == MIXXX_NOSOUNDAPI_STRING) {
```
This breaks existing configurations. 
If this is the user data "None" should be defined as constant as well. Something like kNoApiProgramaticString or such.

ronso0 replied:
I'll use existing `SoundManagerConfig::kDefaultAPI` so that string is defined in one place only
https://github.com/mixxxdj/mixxx/blob/5fe525fe73450039ef4cef2067830705244aed26/src/soundio/soundmanagerconfig.cpp#L14

daschuer commented on src/soundio/soundmanager.cpp:
```diff
@@ -121,7 +121,7 @@ QList<SoundDevicePointer> SoundManager::getDeviceList(
     //qDebug() << "SoundManager::getDeviceList";
 
-    if (filterAPI == "None") {
+    if (filterAPI == MIXXX_NOSOUNDAPI_STRING) {
```
I think we need to keep here the programatic string "None"

ronso0 replied:
I'll use existing `SoundManagerConfig::kDefaultAPI` so that string is defined in one place only
https://github.com/mixxxdj/mixxx/blob/5fe525fe73450039ef4cef2067830705244aed26/src/soundio/soundmanagerconfig.cpp#L14

daschuer reviewed (changes requested):
Some first comments. This needs some brief tests if it effects the stored user data.

ronso0 commented:
I implemented your proposals and rebased (small commits, easy to review).
User data is not touched anymore, just the display strings in Preferences and the effect widgets.

Holzhaus commented on src/effects/effectsmanager.cpp:
```diff
@@ -13,6 +13,8 @@
 #include "moc_effectsmanager.cpp"
 #include "util/assert.h"
 
+const QString EffectsManager::kNoEffectString = QStringLiteral(" --- ");
```
```suggestion
const QString EffectsManager::kNoEffectString = QStringLiteral("---");
```
?

ronso0 replied:
Then it would stick to the left edge (or right edge, depending on how skin desgners align the strings).

Holzhaus replied:
Shouldn't that be fixed by using padding instead of using spaces for alignment?

ronso0 replied:
yes, but the padding would affect all effect names, and I would just like to push the placeholder a bit towards the center, below the other effects.

Holzhaus replied:
IMHO this looks really weird and might lead to bug reports. I'd prefer just "---" without spaces. Can you elaborate why you want to indent it?

![Screenshot from 2021-01-06 14-58-34](https://user-images.githubusercontent.com/1834516/103776575-eaecbb00-502f-11eb-81b5-c10205aed8ce.png)

Holzhaus replied:
Same here:
![Screenshot from 2021-01-06 15-02-19](https://user-images.githubusercontent.com/1834516/103776921-723a2e80-5030-11eb-8298-99cc1c18a567.png)

ronso0 replied:
I doubt it leads to bug reports, but I don't mind. Initially I was just looking at the empyt effect at the bottom of the list and I didn't want it to look like a bullet point. Interestingly that doesn't look good at the top of the list (sound API & devices).

I'll remove the whitespaces.

ronso0 replied:
Done. `---` in effect widgets & prefs

Holzhaus reviewed (commented):
Thanks. I noticed an issue with the effect selector, but that may be unrelated to this branch:

![Peek 2020-12-29 15-22](https://user-images.githubusercontent.com/1834516/103290469-ea2da680-49e9-11eb-9f2d-e4a93a639321.gif)

ronso0 commented:
> Thanks. I noticed an issue with the effect selector, but that may be unrelated to this branch:

That's unrelated.
Actually I only noticed it on Windows until now, because on Linux the menu is drawn over the effect selector (and positioned differently in general it seems).
Didn't yet find a pseudo state to style this...

ronso0 commented:
still ready..

uklotzde commented on src/preferences/dialog/dlgprefeq.cpp:
```diff
@@ -672,11 +672,11 @@ void DlgPrefEQ::setUpMasterEQ() {
     }
     //: Displayed when no effect is selected
-    comboBoxMasterEq->addItem(tr("None"), QVariant());
+    comboBoxMasterEq->addItem(EffectsManager::kNoEffectString); // " --- "
```
Duplicating the actual value of constants in comments is pointless. We don't want to repeat this string literal, not even in comments which would become stale and inconsistent if the value is updated

ronso0 replied:
I removed all duplicates in comments.

uklotzde commented on src/preferences/dialog/dlgprefeq.cpp:
```diff
@@ -672,11 +672,11 @@ void DlgPrefEQ::setUpMasterEQ() {
     int masterEqIndex = comboBoxMasterEq->findData(configuredEffect);
     if (masterEqIndex < 0) {
-        masterEqIndex = availableMasterEQEffects.size(); // selects "None"
+        masterEqIndex = availableMasterEQEffects.size(); // selects " --- "
```
// Discard the current selection

uklotzde commented on src/preferences/dialog/dlgprefeq.cpp:
```diff
@@ -255,9 +255,9 @@ void DlgPrefEQ::slotPopulateDeckEffectSelectors() {
+        box->addItem(EffectsManager::kNoEffectString); // " --- "
         if (selectedEffectId.isEmpty()) {
-            currentIndex = availableQuickEffects.size(); // selects "None"
+            currentIndex = availableQuickEffects.size(); // selects " --- "
```
// Discard the current selection

uklotzde commented on src/preferences/dialog/dlgprefeq.cpp:
```diff
@@ -226,9 +226,9 @@ void DlgPrefEQ::slotPopulateDeckEffectSelectors() {
+        box->addItem(EffectsManager::kNoEffectString); // " --- "
         if (selectedEffectId.isEmpty()) {
-            currentIndex = availableEQEffects.size(); // selects "None"
+            currentIndex = availableEQEffects.size(); // selects " --- "
```
// Discard the current selection

uklotzde reviewed (changes requested):
Please remove the redundant string literals from comments. Instead describe the purpose of the code without repeating parts of the implementation in comments.

ronso0 commented:
Done @uklotzde

uklotzde commented:
Ping @daschuer

daschuer reviewed (changes requested):
Works good now. Only one minor comment.

ronso0 committed:
Pref. EQ, WEffectSelector: update combobox comments

ronso0 commented on src/preferences/dialog/dlgprefeq.cpp:
```diff
@@ -255,9 +256,10 @@ void DlgPrefEQ::slotPopulateDeckEffectSelectors() {
             }
         }
         //: Displayed when no effect is selected
```
aaah that's what they are.
comments updated.

Holzhaus reviewed (approved):
I'm getting this debug assertion on shutdown:

DEBUG ASSERT: "!"Controls were leaked!"" in function void MixxxMainWindow::finalize() at /home/jan/Projects/mixxx/src/mixxx.cpp:870

However, this seems to be a bug in the 2.3 branch, not yours :(

LGTM, thank you.

Be-ing commented:
I don't like this change. This feels like a regression. What was ever the problem with "None"? What was the motivation for changing this? I cannot recall any problems ever caused by the string "None". Who would know what `---` means in this context?

![image](https://user-images.githubusercontent.com/9455094/104143340-aab96f80-5384-11eb-8795-3432cc39f7b2.png)

ronso0 commented:
> ... because in the Sound I/O tabs and in skins' effect regions `---` is much easier to spot/recognize than `None`.
> 
> https://bugs.launchpad.net/mixxx/+bug/1892048

daschuer commented:
I support the --- idea. It is much more recognizable especially if you also consider internationalisation. 
And even in English there is the idiom "to dash something out"
